src/server/JasmineGraphInstanceService.cpp

```diff
@@ -2181,4 +2654,587 @@ void JasmineGraphInstanceService::trainPartition(string trainData){
+/**
+ * Compute the common entities accross all parties given a chainable pairwise common entity information
+ * @param pairwiseCommonEntities Map of party-ids to the mappings of common entities between other parties
+ * @return A map of party ID to a vector of cluster indices that correspond to the common entities
+ */
+map<string, vector<string>>
+synchronizeCommonEntities(map<string, map<string, map<string, string>>> pairwiseCommonEntities) {
+    map<string, vector<string>> partyCommonEntityMap;
+    string firstPassEndParty;
+
+    //Get common entitiy ids of self
+    string currentParty = "A"; //Self party ID
+    string nextParty = pairwiseCommonEntities[currentParty].begin()->first;
+    vector<string> currentPartyIds;
+    for (auto idPairs: pairwiseCommonEntities[currentParty][nextParty]) {
+        currentPartyIds.emplace_back(idPairs.first);
+    }
+
+    //First pass iteration through intermediate pairwise results
+    for (int i = 0; i < pairwiseCommonEntities.size(); i++) {
+        auto commonEntityMap = pairwiseCommonEntities[currentParty][nextParty];
+        //Iterate through common entity ids of currentParty with nextParty
+        vector<string> nextPartyIds;
+        for (string id: currentPartyIds) {
+            partyCommonEntityMap[currentParty].emplace_back(id);
+            //If id is present in the next party common entities, mark it to check in the next iterationa
+            if (commonEntityMap.find(id) != commonEntityMap.end()) {
+                nextPartyIds.emplace_back(pairwiseCommonEntities[currentParty][nextParty][id]);
+            }
+        }
+        //Party to iterate
+        currentParty = nextParty;
+        //Filtered next set of ids, when the loop terminates we will have entities of self which are common for all parties
+        currentPartyIds = nextPartyIds;
+        //Change pointer to next paty in the chain
+        nextParty = pairwiseCommonEntities[currentParty].begin()->first;
+    }
+
+    //Second pass to compute common entities across all entities
+    for (int i = 0; i < pairwiseCommonEntities.size(); i++) {
+        //Replace with the entities of current party which are common for all parties
+        partyCommonEntityMap[currentParty] = currentPartyIds;
+
+        nextParty = pairwiseCommonEntities[currentParty].begin()->first;
+        vector<string> nextPartyIds;
+        //For only the entities common for all, get mapping entities of next party
+        for (string id: currentPartyIds) {
+            nextPartyIds.emplace_back(pairwiseCommonEntities[currentParty][nextParty][id]);
+        }
+        //Party to iterate
+        currentParty = nextParty;
+        //Filtered next set of ids
+        currentPartyIds = nextPartyIds;
+        //Change pointer to next paty in the chain
+        nextParty = pairwiseCommonEntities[currentParty].begin()->first;
+    }
+
+    return partyCommonEntityMap;
+}
+
+void JasmineGraphInstanceService::entityRes(string trainData) {
+
+//    Utils utils;
```

Comment:
We should either uncomment and use these commented code or permanently delete them.